pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,844 @@
+
+	return nil, nil
+}
+
+// calc sha1 hash
+func (h *Handler) sha1(s string) string {
+	hash := sha1.New()
+	hash.Write([]byte(s))
+	sha1sum := hash.Sum(nil)
+	return fmt.Sprintf("%x", sha1sum)
+}
+
+func (h *Handler) checkHash(setting *harvesterv1.Setting) bool {
+	currentHash := h.sha1(setting.Value)
+	savedHash := setting.Annotations[HashStorageNetworkAnnotation]
+	return currentHash == savedHash
+}
+
+func (h *Handler) saveHash(setting *harvesterv1.Setting) {
+	setting.Annotations[HashStorageNetworkAnnotation] = h.sha1(setting.Value)
+}
+
+func (h *Handler) createAndSaveNad(setting *harvesterv1.Setting) error {
+	var config Config
+	bridgeConfig := NewBridgeConfig()
+
+	if err := json.Unmarshal([]byte(setting.Value), &config); err != nil {
+		logrus.Warnf("parsing value error %v", err)
+		return err
+	}
+
+	bridgeConfig.Bridge = config.ClusterNetwork + BridgeSuffix
+	bridgeConfig.IPAM.Range = config.Range
+
+	if config.Vlan == 0 {
+		config.Vlan = DefaultPVID
+	}
+	bridgeConfig.Vlan = config.Vlan
+
+	if len(config.Exclude) > 0 {
+		bridgeConfig.IPAM.Exclude = config.Exclude
+	}
+
+	nadConfig, err := json.Marshal(bridgeConfig)
+	if err != nil {
+		logrus.Warnf("output json error %v", err)
+		return err
+	}
+
+	currentHash := h.sha1(setting.Value)
+	nad := nadv1.NetworkAttachmentDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      StorageNetworkNetAttachDefPrefix + currentHash,
+			Namespace: StorageNetworkNetAttachDefNamespace,
+		},
+	}
+	nad.Annotations = map[string]string{
+		StorageNetworkAnnotation: "true",
+	}
+	nad.Spec.Config = string(nadConfig)
+
+	// create nad
+	if _, err := h.networkAttachmentDefinitions.Create(&nad); err != nil {
+		logrus.Warnf("create net-attach-def failed %v", err)
+		return err
+	}
+
+	// save nad name
+	setting.Annotations[OldNadStorageNetworkAnnotation] = setting.Annotations[NadStorageNetworkAnnotation]
+	setting.Annotations[NadStorageNetworkAnnotation] = nad.Namespace + "/" + nad.Name
+
+	return nil
+}
+
+func (h *Handler) checkChange(setting *harvesterv1.Setting) (bool, error) {
+	if !h.checkHash(setting) {
+		if setting.Value == "" {
+			setting.Annotations[OldNadStorageNetworkAnnotation] = setting.Annotations[NadStorageNetworkAnnotation]
+			setting.Annotations[NadStorageNetworkAnnotation] = ""
+			h.saveHash(setting)
+
+			if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, "enqueue"); err != nil {
+				logrus.Warnf("reset update status error %v", err)
+			}
+			h.settingsController.Enqueue(setting.Name)
+			return false, nil
+		} else {
+			if err := h.createAndSaveNad(setting); err != nil {
+				return false, err
+			}
+
+			h.saveHash(setting)
+
+			if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, "create nad"); err != nil {
+				logrus.Warnf("create nad update status error %v", err)
+			}
+			h.settingsController.Enqueue(setting.Name)
+			return false, nil
+		}
+	}
+	return true, nil
```

Comment:
> it is not a good practice to use enqueue to achieve the reconciliation here and seems we have some duplicate methods with the setting's controller like a hash check, we should consider using a more straightforward approach:

We could not reuse the hash check in the basic setting controller, it will only execute once.
But in this controller, we need to reconcile again and again to wait for the status change because we need to wait between different states.
So we will have lots of states with the same value.

To keep the state clear and independent from other states, we need to separate each stage.

We could not remove old NAD before the Longhorn setting is updated.
Longhorn will still use the old NAD name if some Pods are restarted.

Reply:
I have reviewed somehow a similar PR https://github.com/harvester/harvester/pull/2829/ today; maybe we can take some references from it. Basically, we can have one onChange to do the required operations like scale down the deployment to unmount PVCs, and another one to just check if all volumes are unmounted and ready to update the longhorn setting, we should try to keep the controller's logic more intuitive.

Reply:
Does Wrangler ensure the execute order between multiple onChange function in one controller?
I think it will be a key point.
If Wrangler couldn't guarantee the order, it will still be complex because it still needs to do every necessary check.